dhall/src/Dhall/Tutorial.hs

```diff
@@ -1916,6 +1919,54 @@ import Dhall
 -- * Prefer general purpose built-ins or built-ins appropriate for the task of program configuration
 -- * Design built-ins to catch errors as early as possible (i.e. when type-checking the configuration)
 
+-- $substitutions
+--
+-- Substitutions are another way to extend the language.
+-- Suppose we have the following Haskell datatype:
+--
+-- > data Result = Failure Integer | Success String
+-- >     deriving (Eq, Generic, Show)
+-- >
+-- > instance FromDhall Result
+--
+-- We can use it in Dhall like that:
+--
+-- > -- example.dhall
+-- >
+-- > let Result = < Failure : Integer | Success Text >
+-- > in Result.Failure 1
+--
+-- Right now it is quite easy to keep these two definitions (the one in Haskell source and the one in the Dhall file) synchronized:
+-- If we implement a new feature in the Haskell source we update the corresponding type in the Dhall file.
+-- But what happens if our application is growing and our Result type contains e.g. 10 unions with possible types embedded in it?
+-- Maintaining the code will get tedious. Luckily we can extract the correct Dhall type from the Haskell definition:
+--
+-- > resultDecoder :: Dhall.Decoder Result
+-- > resultDecoder = Dhall.auto
+-- >
+-- > resultType :: Expr Src Void
+-- > resultType = Dhall.expected resultDecoder
+-- >
+-- > resultTypeString :: String
+-- > resultTypeString = show $ pretty resultType
+--
+-- Now we just have to inject that type into the Dhall code and we are done. One common way to do that is to wrap the import of example.dhall in a let expression:
+--
+-- > Dhall.input (Dhall.auto :: Dhall.Decoder Result) ("let Result = " ++ Data.Text.pack resultTypeString ++ " in ./example.dhall")
+--
+-- Now we can omit the definition of Result in our example.dhall file. While this will work perfectly Dhall provides a cleaner solution for our \"injection problem\":
+--
+-- > {-# LANGUAGE OverloadedStrings #-}
+-- >
+-- > myexample :: IO Result
+-- > myexample = let
+-- >    evaluateSettings = Lens.over Dhall.substitutions (Data.Map.insert "Result" resultType) Dhall.defaultEvaluateSettings
```

Comment:
I wonder whether it would be better if the collection of substitutions were opaque instead of a plain `Map`.

Reply:
Since substitutions are a simple mapping and the Map datatype is a common one the user will get an immediate intuition how to construct/update/... the data structure.
I think this contributes more to the usability than an opaque type. Of course it makes it more difficult to change this feature in the future because it might break other peoples code.
I'm not sure about this trade off myself...